lib/Conversion/FSMToSV/FSMToSV.cpp

```diff
@@ -0,0 +1,540 @@
+
+  // Build a SV case-based combinational mux the values provided in
+  // 'stateToValue' to a retured wire.
+  // 'stateToValue' being a list implies that multiple muxes can be emitted at
+  // once, avoiding bloating the IR with a case statement for every muxed value.
+  // A wire is returned for each srcMap provided.
+  // 'nameF' can be provided to specify the name of the output wire created for
+  // each source map.
+  llvm::SmallVector<Value>
+  buildStateCaseMux(Location loc, Value sel,
+                    llvm::ArrayRef<llvm::SmallDenseMap<StateOp, Value>> srcMaps,
+                    llvm::function_ref<StringAttr(size_t)> nameF = {});
+
+  // A handle to the state encoder for this machine.
+  std::unique_ptr<StateEncoding> encoding;
+
+  // A deterministic ordering of the states in this machine.
+  llvm::SmallVector<StateOp> orderedStates;
+
+  // A mapping from a state op to its next-state value.
+  llvm::SmallDenseMap<StateOp, Value> nextStateFromState;
+
+  // A handle to the MachineOp being converted.
+  MachineOp machineOp;
+
+  // A handle to the HW ModuleOp being created.
+  hw::HWModuleOp hwModuleOp;
+
+  // A handle to the state register of the machine.
+  seq::CompRegOp stateReg;
+
+  OpBuilder &b;
+};
+
+FailureOr<Operation *>
+MachineOpConverter::moveOps(Block *block,
+                            llvm::function_ref<bool(Operation *)> exclude) {
+  for (auto &op : llvm::make_early_inc_range(*block)) {
+    if (!isa<comb::CombDialect, hw::HWDialect, fsm::FSMDialect>(
+            op.getDialect()))
+      return op.emitOpError()
+             << "is unsupported (op from the "
+             << op.getDialect()->getNamespace() << " dialect).";
+
+    if (exclude && exclude(&op))
+      continue;
+
+    if (op.hasTrait<OpTrait::IsTerminator>())
+      return &op;
+
+    op.moveBefore(&hwModuleOp.front(), b.getInsertionPoint());
+  }
+  return nullptr;
+}
+
+llvm::SmallVector<Value> MachineOpConverter::buildStateCaseMux(
+    Location loc, Value sel,
+    llvm::ArrayRef<llvm::SmallDenseMap<StateOp, Value>> srcMaps,
+    llvm::function_ref<StringAttr(size_t)> nameF) {
+  sv::CaseOp caseMux;
+  auto caseMuxCtor = [&]() {
+    caseMux = b.create<sv::CaseOp>(loc, CaseStmtType::CaseStmt, sel,
+                                   /*numCases=*/machineOp.getNumStates(),
+                                   [&](size_t caseIdx) {
+                                     StateOp state = orderedStates[caseIdx];
+                                     return encoding->getCasePattern(state);
+                                   });
+  };
+  b.create<sv::AlwaysCombOp>(loc, caseMuxCtor);
+
+  llvm::SmallVector<Value> dsts;
+  // note: cannot use llvm::enumerate, makes the underlying iterator const.
+  size_t idx = 0;
+  for (auto srcMap : srcMaps) {
+    auto valueType = srcMap.begin()->second.getType();
+    StringAttr name;
+    if (nameF)
+      name = nameF(idx);
+
+    auto dst = b.create<sv::RegOp>(loc, valueType, name);
```

Comment:
Why `sv.reg` vs `sv.logic` or `sv.wire`?

Reply:
`sv::BPAssignOp` requires disallows assignments to a net type (`wire`). Seeing as we're not intending to infer a register here, `logic` would indeed be more suitable.